Mathlib/Tactic/Linter/DeprecatedModule.lean

```diff
@@ -0,0 +1,141 @@
+
+to the environment.
+-/
+initialize deprecatedModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := (·.foldl Std.HashSet.insertMany {})
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecatedModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores the `Init` import, since this is a special module that is expected to be imported
+by all files.
+-/
+def addModuleDeprecation {m : Type → Type} [Monad m] [MonadEnv m] [MonadQuotation m] : m Unit := do
+  modifyEnv (deprecatedModuleExt.addEntry ·
+    (← getMainModule, (← getEnv).imports.filterMap fun i =>
+      if i.module == `Init then none else i.module))
+
+/--
+`deprecated_module since yyyy-mm-dd` deprecates the current module `A` in favour of
+its direct imports.
+This means that any file that directly imports `A` will get a notification on the `import A` line
+suggesting to instead import the *direct imports* of `A`.
+-/
+elab (name := deprecated_modules)
+    "deprecated_module " &"since " yyyy:num "-" mm:num "-" dd:num : command => do
+  if yyyy.getNat < 2025 then
+    throwErrorAt yyyy "The year should be at least 2025!"
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02, ..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02, ..., 31!"
+  addModuleDeprecation
+  -- disable the linter, so that it does not complain in the file with the deprecation
+  elabCommand (← `(set_option linter.deprecated.module false))
+
+/-- A utility function to show the pairings `(deprecatedModule, #[preferredModules])`. -/
+elab "#show_deprecated_modules" : command => do
+  let directImports := deprecatedModuleExt.getState (← getEnv)
+  logInfo <| "\n".intercalate <|
+    directImports.fold (init := ["Deprecated modules\n"]) fun nms (i, deps) =>
+      nms ++ [s!"'{i}' deprecates to\n{deps}\n"]
+
+namespace DeprecatedModule
+
+/--
+`IsLaterCommand` is an `IO.Ref` that starts out being `false`.
+As soon as a (non-import) command in a file is processed, the `deprecated.module` linter`
+sets it to `true`.
+If it is `false`, then the `deprecated.module` linter will check for deprecated modules.
+
+This is used to ensure that the linter performs the deprecation checks only once per file.
+There are possible concurrency issues, but they should not be particularly worrying:
+* the linter check should be relatively quick;
+* the only way in which the linter could change what it reports is if the imports are changed
+  and a change in imports triggers a rebuild of the whole file anyway, resetting the `IO.Ref`.
+-/
+initialize IsLaterCommand : IO.Ref Bool ← IO.mkRef false
```

Comment:
I know that this is a little flimsy, but I argue that it is not too bad in the doc-string!

A robust alternative would be to run the linter only on `Parser.isTerminalCommand` (essentially only at `eoi`).

This is guaranteed to run once per file (assuming you do not mess with `... in #exit` in your file), but has the disadvantage that you only get the warning at the very top of the file when parsing has reached the very bottom of the file.